src/realm/dictionary.cpp

```diff
@@ -0,0 +1,252 @@
+/*************************************************************************
+ *
+ * Copyright 2019 Realm Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ **************************************************************************/
+
+#include <realm/dictionary.hpp>
+#include <realm/cluster_tree.hpp>
+#include <realm/array_mixed.hpp>
+#include <realm/group.hpp>
+#include <algorithm>
+
+namespace realm {
+
+DictionaryClusterTree::DictionaryClusterTree(Dictionary* owner, ColumnType col_type, Allocator& alloc, size_t ndx)
+    : ClusterTree(alloc)
+    , m_owner(owner)
+    , m_keys_col(ColKey(ColKey::Idx{0}, col_type, ColumnAttrMask(), 0))
+    , m_ndx_in_cluster(ndx)
+{
+}
+
+DictionaryClusterTree::~DictionaryClusterTree() {}
+
+/******************************** Dictionary *********************************/
+
+Dictionary::Dictionary(const ConstObj& obj, ColKey col_key)
+    : m_obj(obj)
+    , m_col_key(col_key)
+{
+    m_clusters = new DictionaryClusterTree(this, col_key.get_type(), m_obj.get_alloc(), m_obj.get_row_ndx());
+    init_from_parent();
+}
+
+Dictionary::~Dictionary()
+{
+    delete m_clusters;
+}
+
+Dictionary& Dictionary::operator=(const Dictionary& other)
+{
+    if (this != &other) {
+        m_obj = other.m_obj;
+        m_col_key = other.m_col_key;
+
+        if (other.m_clusters) {
+            if (!m_clusters)
+                m_clusters =
+                    new DictionaryClusterTree(this, m_col_key.get_type(), m_obj.get_alloc(), m_obj.get_row_ndx());
+            init_from_parent();
+            m_valid = true;
+        }
+        else {
+            delete m_clusters;
+            m_clusters = nullptr;
+            m_valid = false;
```

Comment:
can we just check if `m_clusters` is null rather than tracking state with another variable?

Reply:
Yes we can. `m_valid` was a leftover from the list implementation. Perhaps it can be changed there as well.